src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.IO.Pipes;
+
+namespace Microsoft.DotNet.Tools.Test;
+
+internal sealed class NamedPipeClient : NamedPipeBase, IClient
+{
+    private readonly NamedPipeClientStream _namedPipeClientStream;
+    private readonly SemaphoreSlim _lock = new(1, 1);
+
+    private readonly MemoryStream _serializationBuffer = new();
+    private readonly MemoryStream _messageBuffer = new();
+    private readonly byte[] _readBuffer = new byte[250000];
+
+    private bool _disposed;
+
+    public NamedPipeClient(string name)
+    {
+        _namedPipeClientStream = new(".", name, PipeDirection.InOut);
+        PipeName = name;
+    }
+
+    public string PipeName { get; }
+
+    public bool IsConnected => _namedPipeClientStream.IsConnected;
+
+    public async Task ConnectAsync(CancellationToken cancellationToken)
+        => await _namedPipeClientStream.ConnectAsync(cancellationToken);
+
+    public async Task<TResponse> RequestReplyAsync<TRequest, TResponse>(TRequest request, CancellationToken cancellationToken)
+       where TRequest : IRequest
+       where TResponse : IResponse
+    {
+        await _lock.WaitAsync(cancellationToken);
+        try
+        {
+            INamedPipeSerializer requestNamedPipeSerializer = GetSerializer(typeof(TRequest));
+
+            // Ask to serialize the body
+            _serializationBuffer.Position = 0;
+            requestNamedPipeSerializer.Serialize(request, _serializationBuffer);
+
+            // Write the message size
+            _messageBuffer.Position = 0;
+
+            // The length of the message is the size of the message plus one byte to store the serializer id
+            // Space for the message
+            int sizeOfTheWholeMessage = (int)_serializationBuffer.Position;
+
+            // Space for the serializer id
+            sizeOfTheWholeMessage += sizeof(int);
+
+            // Write the message size
+#if NETCOREAPP
```

Comment:
Do we need all these pragmas? Is `dotnet` compiled from something else that netcore?

Reply:
Those files are included in the Microsoft.NET.Build.Extensions.Tasks (which uses both net47 and net9) project, because we have a task there that uses them.